File: frontend/projects/irrtirer/src/app/shared/mosaic-generator/events/event-bus.ts
import { Observable, Subject } from 'rxjs';
import { SectorOrderedArgs } from './sector-ordered-args.interface';
import { SectionOrderedArgs } from './section-ordered-args.interface';
import { MosaicSetOrderedArgs } from './mosaic-set-ordered-args.interface';
import { MosaicSetModel } from '../sectors/mosaic-set.model';
import { SectionModel } from '../sectors/section.model';
import { SectorModel } from '../sectors/sector.model';

export class EventBus {
    private readonly mosaicSetOrderingProgress: Subject<MosaicSetOrderedArgs> = new Subject<MosaicSetOrderedArgs>();
    public mosaicSetOrderingProgress$: Observable<MosaicSetOrderedArgs> = this.mosaicSetOrderingProgress.asObservable();

    private readonly sectorOrderingProgress: Subject<SectorOrderedArgs> = new Subject<SectorOrderedArgs>();
    public sectorOrderingProgress$: Observable<SectorOrderedArgs> = this.sectorOrderingProgress.asObservable();

    private readonly sectionOrderingProgress: Subject<SectionOrderedArgs> = new Subject<SectionOrderedArgs>();
    public sectionOrderingProgress$: Observable<SectionOrderedArgs> = this.sectionOrderingProgress.asObservable();

    public emitMosaicSetOrderingEnd(mosaicSet: MosaicSetModel): void {
        this.mosaicSetOrderingProgress.next({ mosaicSet });
    }

    public emitSectorOrderingEnd(sectorModel: SectorModel): void {
        this.sectorOrderingProgress.next({ sector: sectorModel });
    }

    public emitSectionOrderingEnd(sectionModel: SectionModel): void {
        this.sectionOrderingProgress.next({ tiles: sectionModel.directTiles });
    }
}

File: backend/Irrtirer/Irrtirer.Generator/Orderer/EventBus.cs
using Irrtirer.Generator.Models.Events;
using Irrtirer.Generator.MosaicStructure;
using System;
using System.Linq;

namespace Irrtirer.Generator.Orderer
{
    internal class EventBus
    {
        public event EventHandler<MosaicSetOrderedArgs> MosaicSetOrderingProgress;
        public event EventHandler<SectorOrderedArgs> SectorOrderingProgress;
        public event EventHandler<SectionOrderedArgs> SectionOrderingProgress;

        public void OnMosaicSetOrderingEnd(MosaicSetModel mosaicSet)
        {
            MosaicSetOrderingProgress?.Invoke(this, new MosaicSetOrderedArgs()
            {
                Tiles = mosaicSet.MosaicSetTiles.Select(x => x.ToResult())
            });
        }

        public void OnSectorOrderingEnd(SectorModel sectorModel)
        {
            SectorOrderingProgress?.Invoke(this, new SectorOrderedArgs()
            {
                Tiles = sectorModel.SectorTiles.Select(x => x.ToResult())
            });
        }

        public void OnSectionOrderingEnd(SectionModel sectionModel)
        {
            SectionOrderingProgress?.Invoke(this, new SectionOrderedArgs()
            {
                Tiles = sectionModel.DirectTiles.Select(x => x.ToResult())
            });
        }
    }
}
